io.openems.edge.timeofusetariff.entsoe/src/io/openems/edge/timeofusetariff/entsoe/TouEntsoeImpl.java

```diff
@@ -0,0 +1,182 @@
+import io.openems.edge.timeofusetariff.api.TimeOfUseTariff;
+import io.openems.edge.timeofusetariff.api.utils.TimeOfUseTariffUtils;
+
+@Designate(ocd = Config.class, factory = true)
+@Component(//
+		name = "TimeOfUseTariff.ENTSO-E", //
+		immediate = true, //
+		configurationPolicy = ConfigurationPolicy.REQUIRE //
+)
+public class TouEntsoeImpl extends AbstractOpenemsComponent implements TouEntsoe, OpenemsComponent, TimeOfUseTariff {
+
+	private static final int API_EXECUTE_HOUR = 14;
+
+	private final Logger log = LoggerFactory.getLogger(TouEntsoeImpl.class);
+	private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
+	private final AtomicReference<ImmutableSortedMap<ZonedDateTime, Float>> prices = new AtomicReference<>(
+			ImmutableSortedMap.of());
+	private final List<Consumer<Value<Integer>>> subscriptions = new ArrayList<>();
+
+	@Reference
+	private Meta meta;
+
+	private Config config = null;
+	private Currency currency;
+	private ZonedDateTime updateTimeStamp = null;
+
+	public TouEntsoeImpl() {
+		super(//
+				OpenemsComponent.ChannelId.values(), //
+				TouEntsoe.ChannelId.values() //
+		);
+	}
+
+	@Activate
+	private void activate(ComponentContext context, Config config) {
+		super.activate(context, config.id(), config.alias(), config.enabled());
+
+		if (!config.enabled()) {
+			return;
+		}
+		this.config = config;
+
+		Consumer<Currency> updateCurrency = currency -> {
+			this.currency = currency;
+			this.executor.schedule(this.task, 0, TimeUnit.SECONDS);
+		};
+
+		// Trigger to activate when the currency from Meta is updated.
+		this.subscribe(updateCurrency);
+
+		updateCurrency.accept(this.meta.getCurrencyChannel().getNextValue().asEnum());
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		super.deactivate();
+		this.unsubscribeAll();
+		ThreadPoolUtils.shutdownAndAwaitTermination(this.executor, 0);
+	}
+
+	/**
+	 * Subscribes to the Currency channel to trigger on update.
+	 * 
+	 * @param updateCurrency The callback {@link Consumer}.
+	 */
+	private void subscribe(Consumer<Currency> updateCurrency) {
+
+		Consumer<Value<Integer>> subscription = currency -> {
+			updateCurrency.accept(currency.asEnum());
+		};
+		this.subscriptions.add(subscription);
+		this.meta.getCurrencyChannel().onSetNextValue(subscription);
+	}
+
+	/**
+	 * Unsubscribes from all the Subscriptions.
+	 */
+	private void unsubscribeAll() {
+		this.subscriptions.forEach(subscription -> {
+			this.meta.getCurrencyChannel().removeOnSetNextValueCallback(subscription);
+		});
+	}
+
+	private final Runnable task = () -> {
+		var token = this.config.securityToken();
+		var areaCode = this.config.biddingZone().getCode();
+		var fromDate = ZonedDateTime.now().truncatedTo(ChronoUnit.HOURS);
+		var toDate = fromDate.plusDays(1);
+		var unableToUpdatePrices = false;
+
+		try {
+			var result = EntsoeApi.query(token, areaCode, fromDate, toDate);
+
+			final double exchangeRate = this.currency == Currency.EUR //
```

Comment:
ENTSO-E provides the actual Currency of the data. You should use that instead of assuming that it is always EUR.

See `<currency_Unit.name>EUR</currency_Unit.name>`

Reply:
This is not assuming. I'm directly assigning the exchange rate when the user's currency is EUR. While it's possible to retrieve the exchange rate through the API, there are instances where API responses may experience delays. By preemptively handling cases where the currency is EUR, we can avoid unnecessary queries and potential delays.

Reply:
Please read my comment again. You need to compare the currency returned from `EntsoeApi.query()` with the currency configured in `Core.Meta`. None of both is fixed to EUR.

Reply:
Fixed it